### ISSUE TITLE
Phase 4 Hands-On Practice: Transformer 깊이 파기 (Week 15-20)

### ISSUE THREAD

**Issue by SaraHan774** [SaraHan774/26weeks-of-ai#8]
# Phase 4: Transformer 깊이 파기 (Week 15-20)

> **목표**: Transformer를 밑바닥부터 구현하고, GPT/BERT의 차이를 이해하며, 로컬에서 SLM을 실행합니다.

**로드맵 연결**: `llm-learning-roadmap.md` Phase 4 (15-20주)

---

## 📋 Overview

Phase 4에서는 2개의 대형 프로젝트를 진행합니다:

| 주차 | 프로젝트 | 핵심 학습 | 난이도 |
|:---:|:---|:---|:---:|
| Week 15-18 | nanoGPT 완전 정복 | Transformer 구현, Self-Attention, Positional Encoding | ⭐⭐⭐⭐⭐ |
| Week 19-20 | GPT vs BERT 실전 + 로컬 SLM | Decoder/Encoder 차이, 양자화, Ollama | ⭐⭐⭐⭐ |

---

### 목표
Andrej Karpathy의 nanoGPT를 밑바닥부터 구현하고, tiny Shakespeare 데이터셋으로 텍스트를 생성합니다.

### 왜 이 프로젝트인가?
- **Attention is All You Need 실전**: 논문을 코드로 구현
- **GPT 작동 원리**: ChatGPT의 핵심 아키텍처 이해
- **추론 모델 이해**: o1, DeepSeek-R1이 어떻게 "생각"하는지

#### Week 15: Attention is All You Need 논문 + o1 체험
- 논문 정독 (Figure 1 직접 그리며 이해)
- o1으로 복잡한 알고리즘 문제 풀고 사고 로그 분석
- Transformer 구조 전체 이해

#### Week 16: Embedding + Positional Encoding
- Token Embedding 구현
- Positional Encoding 수식 이해 및 구현
- 시각화로 검증

#### Week 17: Multi-Head Attention
- Self-Attention 메커니즘 구현
- Multi-Head의 의미 이해
- Attention Score 시각화

#### Week 18: 전체 Transformer 완성
- Feed-Forward Network
- Layer Normalization, Residual Connection
- 학습 및 텍스트 생성

### 기술 스택
```python
- PyTorch
- matplotlib (시각화)
- tiktoken (토큰화)
```

#### Step 1: 데이터 준비
```python
import torch
import tiktoken

# tiny Shakespeare 다운로드
!wget https://raw.githubusercontent.com/karpathy/char-rnn/master/data/tinyshakespeare/input.txt

# 데이터 로드
with open('input.txt', 'r', encoding='utf-8') as f:
    text = f.read()

print(f"Dataset length: {len(text):,} characters")
print(f"First 100 chars: {text[:100]}")

# 토크나이저 (GPT-2 BPE)
enc = tiktoken.get_encoding("gpt2")
tokens = enc.encode(text)
print(f"Total tokens: {len(tokens):,}")

# Train/Val split
n = int(0.9 * len(tokens))
train_data = torch.tensor(tokens[:n], dtype=torch.long)
val_data = torch.tensor(tokens[n:], dtype=torch.long)

# 배치 생성 함수
def get_batch(split, block_size=256, batch_size=64):
    data = train_data if split == 'train' else val_data
    ix = torch.randint(len(data) - block_size, (batch_size,))
    x = torch.stack([data[i:i+block_size] for i in ix])
    y = torch.stack([data[i+1:i+block_size+1] for i in ix])
    return x, y

x, y = get_batch('train')
print(f"Input shape: {x.shape}, Target shape: {y.shape}")
```

#### Step 2: Positional Encoding
```python
import math
import torch.nn as nn

class PositionalEncoding(nn.Module):
    def __init__(self, d_model, max_len=5000):
        super().__init__()
        
        # PE(pos, 2i) = sin(pos / 10000^(2i/d_model))
        # PE(pos, 2i+1) = cos(pos / 10000^(2i/d_model))
        
        pe = torch.zeros(max_len, d_model)
        position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
        div_term = torch.exp(torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model))
        
        pe[:, 0::2] = torch.sin(position * div_term)
        pe[:, 1::2] = torch.cos(position * div_term)
        
        pe = pe.unsqueeze(0)  # (1, max_len, d_model)
        self.register_buffer('pe', pe)
    
    def forward(self, x):
        # x: (batch, seq_len, d_model)
        return x + self.pe[:, :x.size(1), :]

# 시각화
import matplotlib.pyplot as plt

pe_layer = PositionalEncoding(d_model=128, max_len=100)
pe = pe_layer.pe.squeeze(0).numpy()

plt.figure(figsize=(15, 5))
plt.imshow(pe.T, aspect='auto', cmap='RdBu', vmin=-1, vmax=1)
plt.colorbar()
plt.xlabel('Position')
plt.ylabel('Dimension')
plt.title('Positional Encoding Heatmap')
plt.savefig('positional_encoding.png')
plt.show()
```

#### Step 3: Multi-Head Self-Attention
```python
class MultiHeadAttention(nn.Module):
    def __init__(self, d_model, n_heads, dropout=0.1):
        super().__init__()
        assert d_model % n_heads == 0
        
        self.d_model = d_model
        self.n_heads = n_heads
        self.d_k = d_model // n_heads
        
        # Q, K, V projection layers
        self.W_q = nn.Linear(d_model, d_model)
        self.W_k = nn.Linear(d_model, d_model)
        self.W_v = nn.Linear(d_model, d_model)
        
        # Output projection
        self.W_o = nn.Linear(d_model, d_model)
        
        self.dropout = nn.Dropout(dropout)
        
    def split_heads(self, x, batch_size):
        """(batch, seq_len, d_model) -> (batch, n_heads, seq_len, d_k)"""
        x = x.view(batch_size, -1, self.n_heads, self.d_k)
        return x.transpose(1, 2)
    
    def forward(self, x, mask=None):
        batch_size = x.size(0)
        
        # Linear projections
        Q = self.split_heads(self.W_q(x), batch_size)  # (batch, n_heads, seq_len, d_k)
        K = self.split_heads(self.W_k(x), batch_size)
        V = self.split_heads(self.W_v(x), batch_size)
        
        # Scaled Dot-Product Attention
        # scores = Q @ K^T / sqrt(d_k)
        scores = torch.matmul(Q, K.transpose(-2, -1)) / math.sqrt(self.d_k)
        
        # Masking (for causal attention)
        if mask is not None:
            scores = scores.masked_fill(mask == 0, float('-inf'))
        
        # Softmax
        attn_weights = torch.softmax(scores, dim=-1)
        attn_weights = self.dropout(attn_weights)
        
        # Weighted sum of values
        context = torch.matmul(attn_weights, V)  # (batch, n_heads, seq_len, d_k)
        
        # Concatenate heads
        context = context.transpose(1, 2).contiguous().view(batch_size, -1, self.d_model)
        
        # Output projection
        output = self.W_o(context)
        
        return output, attn_weights

# 테스트
d_model = 512
n_heads = 8
seq_len = 10
batch_size = 2

x = torch.randn(batch_size, seq_len, d_model)

# Causal mask (GPT-style)
mask = torch.tril(torch.ones(seq_len, seq_len)).unsqueeze(0).unsqueeze(0)

attn = MultiHeadAttention(d_model, n_heads)
output, weights = attn(x, mask)

print(f"Input: {x.shape}")
print(f"Output: {output.shape}")
print(f"Attention weights: {weights.shape}")  # (batch, n_heads, seq_len, seq_len)
```

#### Step 4: Transformer Block
```python
class FeedForward(nn.Module):
    def __init__(self, d_model, d_ff=2048, dropout=0.1):
        super().__init__()
        self.linear1 = nn.Linear(d_model, d_ff)
        self.dropout = nn.Dropout(dropout)
        self.linear2 = nn.Linear(d_ff, d_model)
        self.gelu = nn.GELU()
    
    def forward(self, x):
        return self.linear2(self.dropout(self.gelu(self.linear1(x))))

class TransformerBlock(nn.Module):
    def __init__(self, d_model, n_heads, d_ff, dropout=0.1):
        super().__init__()
        self.attn = MultiHeadAttention(d_model, n_heads, dropout)
        self.ff = FeedForward(d_model, d_ff, dropout)
        self.ln1 = nn.LayerNorm(d_model)
        self.ln2 = nn.LayerNorm(d_model)
        self.dropout = nn.Dropout(dropout)
    
    def forward(self, x, mask):
        # Multi-Head Attention + Residual + LayerNorm
        attn_out, _ = self.attn(self.ln1(x), mask)
        x = x + self.dropout(attn_out)
        
        # Feed-Forward + Residual + LayerNorm
        ff_out = self.ff(self.ln2(x))
        x = x + self.dropout(ff_out)
        
        return x
```

#### Step 5: GPT 모델 전체
```python
class GPT(nn.Module):
    def __init__(self, vocab_size, d_model=512, n_heads=8, n_layers=6, d_ff=2048, 
                 max_len=512, dropout=0.1):
        super().__init__()
        
        self.token_embedding = nn.Embedding(vocab_size, d_model)
        self.pos_encoding = PositionalEncoding(d_model, max_len)
        
        self.blocks = nn.ModuleList([
            TransformerBlock(d_model, n_heads, d_ff, dropout)
            for _ in range(n_layers)
        ])
        
        self.ln_f = nn.LayerNorm(d_model)
        self.head = nn.Linear(d_model, vocab_size, bias=False)
        
        # Weight tying
        self.head.weight = self.token_embedding.weight
        
        self.apply(self._init_weights)
    
    def _init_weights(self, module):
        if isinstance(module, nn.Linear):
            torch.nn.init.normal_(module.weight, mean=0.0, std=0.02)
            if module.bias is not None:
                torch.nn.init.zeros_(module.bias)
        elif isinstance(module, nn.Embedding):
            torch.nn.init.normal_(module.weight, mean=0.0, std=0.02)
    
    def forward(self, idx, targets=None):
        B, T = idx.shape
        
        # Token + Positional Embeddings
        tok_emb = self.token_embedding(idx)  # (B, T, d_model)
        x = self.pos_encoding(tok_emb)
        
        # Causal mask
        mask = torch.tril(torch.ones(T, T, device=idx.device)).unsqueeze(0).unsqueeze(0)
        
        # Transformer blocks
        for block in self.blocks:
            x = block(x, mask)
        
        x = self.ln_f(x)
        logits = self.head(x)  # (B, T, vocab_size)
        
        # Loss calculation
        if targets is not None:
            loss = nn.functional.cross_entropy(logits.view(-1, logits.size(-1)), targets.view(-1))
            return logits, loss
        
        return logits

# 모델 초기화
vocab_size = 50257  # GPT-2 vocab size
model = GPT(vocab_size, d_model=384, n_heads=6, n_layers=6, d_ff=1536, max_len=256)

total_params = sum(p.numel() for p in model.parameters())
print(f"Total parameters: {total_params:,}")  # ~30M
```

#### Step 6: 학습
```python
import torch.optim as optim

device = 'cuda' if torch.cuda.is_available() else 'cpu'
model = model.to(device)

optimizer = optim.AdamW(model.parameters(), lr=3e-4)

def train_step():
    model.train()
    x, y = get_batch('train')
    x, y = x.to(device), y.to(device)
    
    logits, loss = model(x, y)
    
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    
    return loss.item()

# 학습 루프
max_iters = 5000
eval_interval = 500

for iter in range(max_iters):
    loss = train_step()
    
    if iter % eval_interval == 0:
        print(f"Iter {iter}: Loss = {loss:.4f}")

torch.save(model.state_dict(), 'gpt_shakespeare.pth')
```

#### Step 7: 텍스트 생성
```python
def generate(model, prompt, max_new_tokens=100, temperature=1.0):
    model.eval()
    
    # Encode prompt
    enc = tiktoken.get_encoding("gpt2")
    idx = torch.tensor(enc.encode(prompt), dtype=torch.long).unsqueeze(0).to(device)
    
    for _ in range(max_new_tokens):
        # Crop context if needed
        idx_cond = idx if idx.size(1) <= 256 else idx[:, -256:]
        
        # Get predictions
        with torch.no_grad():
            logits = model(idx_cond)
        
        # Focus on last time step
        logits = logits[:, -1, :] / temperature
        
        # Sample from distribution
        probs = torch.softmax(logits, dim=-1)
        idx_next = torch.multinomial(probs, num_samples=1)
        
        # Append to sequence
        idx = torch.cat((idx, idx_next), dim=1)
    
    # Decode
    generated = enc.decode(idx[0].tolist())
    return generated

# 생성 테스트
prompt = "To be or not to be"
generated_text = generate(model, prompt, max_new_tokens=200, temperature=0.8)
print(generated_text)
```

#### Attention Score의 의미
```python
# Q @ K^T: 각 토큰이 다른 토큰과 얼마나 "관련" 있는지
# Softmax: 확률 분포로 변환
# @ V: 관련도에 따라 value를 가중 평균
```

#### Temperature의 역할
- `temperature = 1.0`: 원래 분포
- `temperature > 1`: 더 랜덤 (창의적)
- `temperature < 1`: 더 확정적 (보수적)

#### Layer Normalization vs Batch Normalization
- **Batch Norm**: 배치 차원에서 정규화 (이미지에 적합)
- **Layer Norm**: Feature 차원에서 정규화 (시퀀스에 적합)

### 완료 기준
- [ ] Positional Encoding 구현 및 시각화
- [ ] Multi-Head Attention 구현
- [ ] Transformer Block 완성
- [ ] tiny Shakespeare 학습 (Loss < 2.0)
- [ ] 텍스트 생성 (셰익스피어 스타일)

---

### 목표
GPT-2와 BERT를 실전에서 비교하고, Ollama로 Llama 3를 로컬에서 실행합니다.

#### Week 19: GPT vs BERT 비교
1. **GPT-2 (Decoder-only)**
   - 텍스트 생성 (왼→오)
   - Few-shot prompting

2. **BERT (Encoder-only)**
   - 빈칸 채우기 (Masked Language Modeling)
   - 문장 분류

3. **비교 실험**
   - 같은 작업에서 성능 비교
   - 속도, 메모리 사용량 측정

#### Week 20: 로컬 SLM 실행
1. **Ollama 설치 및 Llama 3 실행**
   - 양자화 (4-bit, 8-bit) 비교
   - 추론 속도 측정

2. **"Transformer 작동 원리" 발표 자료 작성**
   - 10분 분량
   - 시각 자료 포함

#### GPT-2 사용법
```python
from transformers import GPT2LMHeadModel, GPT2Tokenizer

# 모델 로드
model_name = 'gpt2'
tokenizer = GPT2Tokenizer.from_pretrained(model_name)
model = GPT2LMHeadModel.from_pretrained(model_name)

# 텍스트 생성
prompt = "Once upon a time"
input_ids = tokenizer.encode(prompt, return_tensors='pt')

output = model.generate(input_ids, max_length=100, temperature=0.8, do_sample=True)
generated_text = tokenizer.decode(output[0], skip_special_tokens=True)

print(generated_text)
```

#### BERT 사용법
```python
from transformers import BertForMaskedLM, BertTokenizer

# 모델 로드
model_name = 'bert-base-uncased'
tokenizer = BertTokenizer.from_pretrained(model_name)
model = BertForMaskedLM.from_pretrained(model_name)

# Masked Language Modeling
text = "The capital of France is [MASK]."
input_ids = tokenizer.encode(text, return_tensors='pt')

# 예측
with torch.no_grad():
    outputs = model(input_ids)
    predictions = outputs.logits

# [MASK] 위치 찾기
mask_token_index = (input_ids == tokenizer.mask_token_id).nonzero(as_tuple=True)[1]

# Top-k 예측
predicted_token_ids = predictions[0, mask_token_index].topk(5).indices[0]
predicted_tokens = [tokenizer.decode([token_id]) for token_id in predicted_token_ids]

print(f"Top 5 predictions: {predicted_tokens}")
```

#### Ollama 실행
```bash
# 설치 (macOS/Linux)
curl -fsSL https://ollama.com/install.sh | sh

# Llama 3 다운로드 (8B 모델)
ollama pull llama3.2:3b

# 실행
ollama run llama3.2:3b

# Python에서 사용
pip install ollama
```

```python
import ollama

# 텍스트 생성
response = ollama.chat(model='llama3.2:3b', messages=[
    {'role': 'user', 'content': 'Explain Transformer in simple terms.'}
])

print(response['message']['content'])

# 양자화 비교
models = ['llama3.2:3b', 'llama3.2:3b-q4_0', 'llama3.2:3b-q8_0']

for model in models:
    import time
    start = time.time()
    response = ollama.chat(model=model, messages=[
        {'role': 'user', 'content': 'Hello!'}
    ])
    elapsed = time.time() - start
    print(f"{model}: {elapsed:.2f}s")
```

### 완료 기준
- [ ] GPT-2와 BERT 실전 비교
- [ ] Ollama로 Llama 3 실행
- [ ] 양자화 성능 비교
- [ ] "Transformer 작동 원리" 발표 자료 완성

---

### 필수 자료
1. **Andrej Karpathy - Let's build GPT**
   - https://www.youtube.com/watch?v=kCc8FmEb1nY
   - Week 16-18: 메인 가이드

2. **The Annotated Transformer**
   - http://nlp.seas.harvard.edu/annotated-transformer/
   - Week 16: 코드 참고

3. **Attention is All You Need**
   - https://arxiv.org/abs/1706.03762
   - Week 15: 필독 논문

4. **Jay Alammar - The Illustrated Transformer**
   - https://jalammar.github.io/illustrated-transformer/
   - Week 15: 시각적 이해

---

**다음 단계**: Phase 5에서 RAG, Multi-Agent, MCP를 구축합니다!

**질문이 있다면 코멘트로 남겨주세요!** 🚀
